File: onboarder/src/main/java/com/chyrta/onboarder/OnboarderPage.java
package com.chyrta.onboarder;

import android.graphics.drawable.Drawable;
import android.support.annotation.ColorRes;
import android.support.annotation.DrawableRes;
import android.support.annotation.StringRes;

public class OnboarderPage {

    public String title;
    public String description;
    public Drawable imageResource;
    @StringRes public int titleResourceId;
    @StringRes public int descriptionResourceId;
    @DrawableRes public int imageResourceId;
    @ColorRes public int titleColor;
    @ColorRes public int descriptionColor;
    @ColorRes public int backgroundColor;

    public String getTitle() {
        return title;
    }

    public int getTitleResourceId() {
        return titleResourceId;
    }

    public String getDescription() {
        return description;
    }

    public int getDescriptionResourceId() {
        return descriptionResourceId;
    }

    public int getTitleColor() {
        return titleColor;
    }

    public int getDescriptionColor() {
        return descriptionColor;
    }

    public int getImageResourceId() {
        return imageResourceId;
    }

    public int getBackgroundColor() {
        return backgroundColor;
    }

    public static Creator newCreator() {
        return new OnboarderPage().new Creator();
    }

    public class Creator {

        private Creator() {}

        public Creator title(String title) {
            OnboarderPage.this.title = title;
            return this;
        }

        public Creator titleResourceId(@StringRes int titleResourceId) {
            OnboarderPage.this.titleResourceId = titleResourceId;
            return this;
        }

        public Creator description(String description) {
            OnboarderPage.this.description = description;
            return this;
        }

        public Creator descriptionResourceId(@StringRes int descriptionResourceId) {
            OnboarderPage.this.descriptionResourceId = descriptionResourceId;
            return this;
        }

        public Creator imageDrawable(Drawable imageDrawable) {
            OnboarderPage.this.imageResource = imageDrawable;
            return this;
        }

        public Creator imageResourceId(@DrawableRes int imageResourceId) {
            OnboarderPage.this.imageResourceId = imageResourceId;
            return this;
        }

        public Creator titleColor(@ColorRes int titleColor) {
            OnboarderPage.this.titleColor = titleColor;
            return this;
        }

        public Creator descriptionColor(@ColorRes int descriptionColor) {
            OnboarderPage.this.descriptionColor = descriptionColor;
            return this;
        }

        public Creator backgroundColor(@ColorRes int backgroundColor) {
            OnboarderPage.this.backgroundColor = backgroundColor;
            return this;
        }

        public OnboarderPage create() {
            return OnboarderPage.this;
        }

    }

}

File: onboarder/src/main/java/com/chyrta/onboarder/OnboarderPage.kt
package com.chyrta.onboarder

import android.graphics.drawable.Drawable
import android.support.annotation.ColorRes
import android.support.annotation.DrawableRes
import android.support.annotation.StringRes

class OnboarderPage {

    var title: String? = null
        private set
    var description: String? = null
        private set
    var imageResource: Drawable? = null
        private set

    var titleResourceId: Int = 0
        @StringRes get() = field
        private set(@StringRes titleResourceId) { field = titleResourceId}

    var descriptionResourceId: Int = 0
        @StringRes get() = field
        private set(@StringRes descriptionResourceId) { field = descriptionResourceId }

    var imageResourceId: Int = 0
        @DrawableRes get() = field
        private set(@DrawableRes value) { field = value }

    var titleColor: Int = 0
        @ColorRes get() = field
        private set(@ColorRes titleColor) { field = titleColor}


    var descriptionColor: Int = 0
        @ColorRes get() = field
        private set(@ColorRes descriptionColor) { field = descriptionColor }

    var backgroundColor: Int = 0
        @ColorRes get() = field
        private set(@ColorRes backgroundColor ) { field = backgroundColor }

    inner class Creator internal constructor() {

        fun title(title: String): Creator {
            this@OnboarderPage.title = title
            return this
        }

        fun titleResourceId(@StringRes titleResourceId: Int): Creator {
            this@OnboarderPage.titleResourceId = titleResourceId
            return this
        }

        fun description(description: String): Creator {
            this@OnboarderPage.description = description
            return this
        }

        fun descriptionResourceId(@StringRes descriptionResourceId: Int): Creator {
            this@OnboarderPage.descriptionResourceId = descriptionResourceId
            return this
        }

        fun imageDrawable(imageDrawable: Drawable): Creator {
            this@OnboarderPage.imageResource = imageDrawable
            return this
        }

        fun imageResourceId(@DrawableRes imageResourceId: Int): Creator {
            this@OnboarderPage.imageResourceId = imageResourceId
            return this
        }

        fun titleColor(@ColorRes titleColor: Int): Creator {
            this@OnboarderPage.titleColor = titleColor
            return this
        }

        fun descriptionColor(@ColorRes descriptionColor: Int): Creator {
            this@OnboarderPage.descriptionColor = descriptionColor
            return this
        }

        fun backgroundColor(@ColorRes backgroundColor: Int): Creator {
            this@OnboarderPage.backgroundColor = backgroundColor
            return this
        }

        fun create(): OnboarderPage {
            return this@OnboarderPage
        }

    }

    companion object {

        fun newCreator(): Creator {
            return OnboarderPage().Creator()
        }
    }

}
